go/exosocket/index.go

```diff
@@ -0,0 +1,106 @@
+package exosocket
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Originate/exocom/go/structs"
+	"github.com/Originate/exocom/go/utils"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+)
+
+// ExoSocket is the low level API to Exocom
+type ExoSocket struct {
+	config                      Config
+	shouldReconnectOnDisconnect bool
+	socket                      *websocket.Conn
+	messageChannel              chan structs.Message
+}
+
+// NewExoSocket returns an ExoSocket with the given config
+func NewExoSocket(config Config) *ExoSocket {
+	return &ExoSocket{config: config}
+}
+
+// Connect brings an ExoSocket instance online
+func (e *ExoSocket) Connect() error {
+	if e.config.Port == 0 {
+		e.config.Port = 80
+	}
+	exocomURL := fmt.Sprintf("ws://%s:%d/services", e.config.Host, e.config.Port)
+	socket, err := utils.ConnectWithRetry(exocomURL, 100)
+	if err != nil {
+		return err
+	}
+	e.shouldReconnectOnDisconnect = true
+	e.socket = socket
+	e.messageChannel = make(chan structs.Message)
+	go e.listenForMessages()
+	_, err = e.Send(MessageOptions{
+		Name: "exocom.register-service",
+	})
+	return err
+}
+
+// Close takes this ExoRelay instance offline
+func (e *ExoSocket) Close() error {
+	e.shouldReconnectOnDisconnect = false
```

Comment:
This looks like it may be called concurrently. If so, this needs to be protected by a mutex or manipulated using the `atomic` package to avoid a data race